File: src/main/java/org/tobi29/scapes/chunk/terrain/infinite/TerrainInfiniteChunkClient.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.chunk.terrain.infinite;

import org.tobi29.scapes.engine.utils.io.tag.TagStructure;
import org.tobi29.scapes.engine.utils.math.vector.Vector2i;
import org.tobi29.scapes.entity.client.EntityClient;
import org.tobi29.scapes.packets.PacketRequestChunk;

import java.util.Optional;

public class TerrainInfiniteChunkClient extends TerrainInfiniteChunk {
    private final Optional<TerrainInfiniteChunkClient> optional =
            Optional.of(this);
    private final TerrainInfiniteClient terrain;
    private final TerrainInfiniteRendererChunk rendererChunk;
    private boolean requested;

    public TerrainInfiniteChunkClient(Vector2i pos,
            TerrainInfiniteClient terrain, int zSize,
            TerrainInfiniteRenderer renderer) {
        super(pos, terrain, terrain.world(), zSize);
        this.terrain = terrain;
        rendererChunk = new TerrainInfiniteRendererChunk(this, renderer);
    }

    public Optional<TerrainInfiniteChunkClient> optional() {
        return optional;
    }

    public void updateClient() {
        if (state.id < State.LOADED.id) {
            TerrainInfiniteClient terrainClient = terrain;
            if (!requested && terrainClient.requestedChunks() < 3) {
                requested = true;
                terrainClient.changeRequestedChunks(1);
                terrain.world().connection()
                        .send(new PacketRequestChunk(pos.intX(), pos.intY()));
            }
        }
    }

    public void setLoaded() {
        state = State.LOADED;
    }

    @Override
    public void update(int x, int y, int z, boolean updateTile) {
        terrain.renderer()
                .blockChange(x + posBlock.intX(), y + posBlock.intY(), z);
        if (state.id >= State.LOADED.id) {
            terrain.lighting()
                    .updateLight(x + posBlock.intX(), y + posBlock.intY(), z);
        }
    }

    @Override
    public void updateLight(int x, int y, int z) {
        terrain.renderer()
                .blockChange(x + posBlock.intX(), y + posBlock.intY(), z);
    }

    public TerrainInfiniteRendererChunk rendererChunk() {
        return rendererChunk;
    }

    public void resetRequest() {
        requested = false;
    }

    public void load(TagStructure tagStructure) {
        bID.load(tagStructure.getList("BlockID"));
        bData.load(tagStructure.getList("BlockData"));
        bLight.load(tagStructure.getList("BlockLight"));
        initHeightMap();
        for (TagStructure tag : tagStructure.getList("Entities")) {
            EntityClient entity =
                    EntityClient.make(tag.getInteger("ID"), terrain.world());
            entity.read(tag.getStructure("Data"));
            terrain.world().addEntity(entity, tag.getInteger("EntityID"));
        }
        metaData = tagStructure.getStructure("MetaData");
        initHeightMap();
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/chunk/terrain/infinite/TerrainInfiniteChunkClient.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.chunk.terrain.infinite

import org.tobi29.scapes.engine.utils.andNull
import org.tobi29.scapes.engine.math.vector.Vector2i
import org.tobi29.scapes.engine.utils.profiler.profilerSection
import org.tobi29.scapes.engine.utils.tag.*
import org.tobi29.scapes.entity.client.EntityClient
import org.tobi29.scapes.entity.client.MobClient
import org.tobi29.scapes.terrain.infinite.TerrainInfiniteBaseChunk

class TerrainInfiniteChunkClient(pos: Vector2i,
                                 val terrain: TerrainInfiniteClient,
                                 zSize: Int,
                                 renderer: TerrainInfiniteRenderer) : TerrainInfiniteChunk<EntityClient>(
        pos, terrain, zSize) {
    private val rendererChunk = TerrainInfiniteRendererChunk(this, renderer)

    fun setLoaded() {
        state = TerrainInfiniteBaseChunk.State.LOADED
    }

    fun updateClient(delta: Double) {
        profilerSection("Entities") {
            entitiesMut.values.forEach { entity ->
                entity.update(delta)
                if (entity is MobClient) {
                    entity.move(delta)
                }
                val pos = entity.getCurrentPos()
                val x = pos.intX() shr 4
                val y = pos.intY() shr 4
                if ((x != this.pos.x || y != this.pos.y) && unmapEntity(
                        entity)) {
                    if (terrain.chunkC(x, y, { chunk ->
                        chunk.mapEntity(entity)
                        true
                    }) == null) {
                        terrain.entityRemoved(entity)
                    }
                }
            }
        }
    }

    override fun dispose() {
        entitiesMut.values.forEach { terrain.entityRemoved(it) }
    }

    override fun update(x: Int,
                        y: Int,
                        z: Int,
                        updateTile: Boolean) {
        terrain.renderer.blockChange(x + posBlock.x, y + posBlock.y, z)
        if (state.id >= TerrainInfiniteBaseChunk.State.LOADED.id) {
            terrain.lighting().updateLight(x + posBlock.x, y + posBlock.y, z)
        }
    }

    override fun updateLight(x: Int,
                             y: Int,
                             z: Int) {
        terrain.renderer.blockChange(x + posBlock.x,
                y + posBlock.y, z)
    }

    fun rendererChunk(): TerrainInfiniteRendererChunk {
        return rendererChunk
    }

    fun read(map: TagMap) {
        val data = data
        map["BlockID"]?.toList()?.let {
            (data.idData.asSequence() zip it.asSequence().mapNotNull(
                    Tag::toMap).andNull()).forEach { (data, tag) ->
                data.read(tag)
            }
        }
        map["BlockData"]?.toList()?.let {
            (data.dataData.asSequence() zip it.asSequence().mapNotNull(
                    Tag::toMap).andNull()).forEach { (data, tag) ->
                data.read(tag)
            }
        }
        map["BlockLight"]?.toList()?.let {
            (data.lightData.asSequence() zip it.asSequence().mapNotNull(
                    Tag::toMap).andNull()).forEach { (data, tag) ->
                data.read(tag)
            }
        }
        map["MetaData"]?.toMap()?.let { metaData = it.toMutTag() }
        initHeightMap()
    }
}
